packages/core/source-map/src/SourceMap.js

```diff
@@ -0,0 +1,441 @@
+// @flow
+import {SourceMapConsumer, SourceMapGenerator} from 'source-map';
+import lineCounter from '../../utils/src/lineCounter';
+
+type PositionType = {
+  line: number,
+  column: number
+};
+
+type MappingType = {
+  generated: PositionType,
+  original: PositionType | null,
+  source: string | null,
+  name?: string | null
+};
+
+type RawSourceMapType = {
+  version: number,
+  sources: Array<string>,
+  names: Array<string>,
+  sourceRoot?: string,
+  sourcesContent?: Array<string>,
+  mappings: string,
+  file: string
+};
+
+type RawMapInputType = SourceMapConsumer | string | RawSourceMapType;
+
+type ConsumerMappingItemType = {
+  source: string,
+  generatedLine: number,
+  generatedColumn: number,
+  originalLine: number,
+  originalColumn: number,
+  name: string
+};
+
+export default class SourceMap {
+  mappings: Array<MappingType>;
+  sources: Map<string, string | null>;
+  lineCount: number;
+
+  constructor(
+    mappings?: Array<MappingType> = [],
+    sources?: Map<string, string> | {[key: string]: string}
+  ) {
+    // We probably only wanna run this in dev, as sourcemaps should be fast in prod...
+    this._validateMappings(mappings);
+
+    this.mappings = mappings;
+
+    if (sources) {
+      let iteratable =
+        typeof sources === 'object' ? Object.entries(sources) : sources;
+      // $FlowFixMe
+      this.sources = new Map(iteratable);
+    } else {
+      this.sources = new Map();
+    }
+
+    this.lineCount = 0;
+  }
+
+  // This should never be necessary as long as sourcemaps are valid
+  // It's still here in case some asset needs it but should be deprecated over time...
+  filterMappings(mappings: Array<MappingType>): Array<MappingType> {
+    return mappings.filter((mapping: MappingType) => {
+      if (!mapping || !mapping.generated || !mapping.source) return false;
+
+      let isValidOriginal =
+        mapping.original === null ||
+        (typeof mapping.original.line === 'number' &&
+          mapping.original.line > 0 &&
+          typeof mapping.original.column === 'number');
+
+      let isValidGenerated =
+        typeof mapping.generated.line === 'number' &&
+        mapping.generated.line > 0 &&
+        typeof mapping.generated.column === 'number';
+
+      return isValidOriginal && isValidGenerated;
+    });
+  }
+
+  _validateMappings(mappings: Array<MappingType>) {
```

Comment:
This should be removed eventually this is just for testing some stuff and seeing if there are invalid mappings somewhere.

Reply:
Since this doesn't rely on `this`, maybe it can be extracted into its own module and imported here instead?

Reply:
sure, would make sense